feat(map): add clustering and pin bitmap generation utilities

M-e-d-A opened the pull request:
# Description:
## Purpose of the change
This PR introduces the core utility classes for map event clustering, as defined in sub-issue #512.

It includes:
* `MapClustering` utility with Union-Find algorithm and Haversine distance calculation for grouping nearby events
* `PinBitmapGenerator` for creating cluster annotation images with event counts
* Comprehensive unit tests for both clustering logic and bitmap generation
* Support for same-location event stacking and zoom-based clustering

## Related issues
Closes #512 
Part of #511 

## How to test
1. Check out the branch.
2. Run the tests in `MapClusteringTest.kt` and `PinBitmapGeneratorTest.kt` via Android Studio or `./gradlew test`.
3. Verify all tests pass.
4. Review the clustering algorithm and bitmap generation logic for correctness.

**Note:** An LLM was used for the sole and exclusive purpose of improving the clarity and formatting of the PR description.

**Related PRs in this series:**
- **#514** *(this PR)* 
- #515

**Screenshots of the final product:**

<table>
  <tr>
    <td align="center">
      <img src="https://github.com/user-attachments/assets/0638e3eb-070d-4bf4-b6ab-449ad68e2cfc" alt="Four events on the exact same location A" width="300px">
      <br>
      <em>Four events on the exact same location A</em>
    </td>
    <td align="center">
      <img src="https://github.com/user-attachments/assets/fb020256-8a0f-4b7f-a480-4fa5f6098d7f" alt="Map with clustering A" width="300px">
      <br>
      <em>Map with clustering A</em>
    </td>
  </tr>
  <tr>
    <td align="center">
      <img src="https://github.com/user-attachments/assets/a300e5e1-e60e-4c5e-bcb2-d4408287c632" alt="Map with clustering B" width="300px">
      <br>
      <em>Map with clustering B</em>
    </td>
    <td align="center">
      <img src="https://github.com/user-attachments/assets/878660b9-82c3-4281-824e-4801d2f5de65" alt="Four events on the exact same location B" width="300px">
      <br>
      <em>Four events on the exact same location B</em>
    </td>
  </tr>
</table>

M-e-d-A committed:
feat(map): add MapClustering utility for event grouping

Implemented clustering logic using Union-Find and Haversine distance.
The utility can stack events at the same location and cluster nearby events based on zoom level.

M-e-d-A committed:
feat(map): add PinBitmapGenerator for cluster annotation images

Creates circular bitmaps with event count for clusters. Handles large numbers with "99+" display.
Uses OnePass purple theme color and proper scaling for various screen densities.

M-e-d-A committed:
test(map): add unit tests for clustering and bitmap generation

Covers stackSameLocationEvents, clusterItemsForZoom, and cluster bitmap creation.
Tests edge cases including same location events, high/low zoom clustering.

RD418 commented on app/src/main/java/ch/onepass/onepass/ui/map/MapClustering.kt:
Great code ! Really, your code is clean, well-documented. And super idea to use union-find for clusters.

M-e-d-A replied:
Thank you!

RD418 commented on app/src/test/java/ch/onepass/onepass/ui/map/MapClusteringTest.kt:
Your tests looks ok to me great job !  Your line coverage is 98% !

RD418 reviewed (changes requested):
### Overall
Your code is super clear and documented. Excellent code coverage, but I have to request change on your forgotten hard-coded color, please check below !

M-e-d-A committed:
chore(map): align cluster pin colors with theme

Updated PinBitmapGenerator to use theme colors (Primary and OnBackground)
instead of hard-coded values. This ensures cluster annotations maintain
visual consistency with the overall app theme and support dynamic theming.

M-e-d-A commented on app/src/main/java/ch/onepass/onepass/ui/map/PinBitmapGenerator.kt:
```diff
@@ -0,0 +1,57 @@
+    val radius = size / 2.5f
+
+    // Draw Circle Background (OnePass Purple)
+    paint.color = Color.parseColor("#6200EE") // Replace with your theme color
```
Thanks for spotting it, fixed!

M-e-d-A commented on app/src/main/java/ch/onepass/onepass/ui/map/PinBitmapGenerator.kt:
```diff
@@ -0,0 +1,57 @@
+    paint.color = Color.WHITE
+    paint.textAlign = Paint.Align.CENTER
+    paint.textSize = 24f * scale
+    paint.typeface = Typeface.create(Typeface.DEFAULT, Typeface.BOLD)
```
Thanks for the remark! Resolved!

RD418 commented on app/src/main/java/ch/onepass/onepass/ui/map/PinBitmapGenerator.kt:
You fixed the requested changes ! GJ.

M-e-d-A replied:
Thanks!

RD418 reviewed (approved):
### Update after RC
Everything is fixed, good job.
